### PR TITLE
draft: add option xmlMode to parse xml with htmlparser2

### LINKED ISSUE #109: "rehype is not an XML parser"
### Initial checklist

- [X] I read the [support docs](https://github.com/rehypejs/.github/blob/main/support.md)
- [X] I read the [contributing guide](https://github.com/rehypejs/.github/blob/main/contributing.md)
- [X] I agree to follow the [code of conduct](https://github.com/rehypejs/.github/blob/main/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Arehypejs&type=Issues) and couldn’t find anything (or linked relevant results below)

### Problem

readme says "rehype is not an XML parser"
but does not help me to find an XML parser for unified

https://github.com/rehypejs/rehype/blob/620bb9cdf53a4dbe70adfca8d66c8824eb3f513f/packages/rehype-parse/readme.md?plain=1#L137-L140

### Solution

suggest an XML parser for unified

something based on [xast-util-from-xml](https://github.com/syntax-tree/xast-util-from-xml)

### Alternatives

`rehype-parse` works for simple XML files

but it fails to parse `<![CDATA[ ... ]]>`

example: [nixpkgs/doc/functions/library/attrsets.xml](https://github.com/NixOS/nixpkgs/blob/10586701d93c9654826e04f43360f8b6e219df35/doc/functions/library/attrsets.xml#L52) (docbook xml format) (https://github.com/NixOS/nixpkgs/issues/105243)

```xml
  <example xml:id="function-library-lib.attrset.attrByPath-example-value-exists">
   <title>Extracting a value from a nested attribute set</title>
<programlisting><![CDATA[
let set = { a = { b = 3; }; };
in lib.attrsets.attrByPath [ "a" "b" ] 0 set
=> 3
]]></programlisting>
  </example>
```

#### Workaround

```js
const inputText = (
  readFileSync(inputPath, 'utf8')
  // workaround for parsing xml
  // https://github.com/rehypejs/rehype/issues/109
  //.replace(/<!\[CDATA\[(.*?)\]\]>/sg, '$1')
  .replace(/<!\[CDATA\[(.*?)\]\]>/sg, '<cdata>$1</cdata>')
);
```

### PR DESCRIPTION
**Pull request by milahu**
add xml parser

fix #109

### Initial checklist

* [ ] I read the support docs 
* [ ] I read the contributing guide 
* [ ] I agree to follow the code of conduct 
* [ ] I searched issues and couldn’t find anything (or linked relevant results below) 
* [ ] If applicable, I’ve added docs and tests

### Description of changes

TODO

### CODE REVIEW

File: packages/rehype-parse/package.json [hunk 1]
```diff
@@ -37,7 +37,9 @@
   ],
   "dependencies": {
     "@types/hast": "^2.0.0",
+    "hast-util-from-htmlparser2": "github:milahu/hast-util-from-htmlparser2",
```
Comment (milahu) [inline packages/rehype-parse/package.json:40]: todo: move to https://github.com/syntax-tree/hast-util-from-htmlparser2

**Comment by wooorm** [comment]
@milahu Hi!

Can you pause for a sec, and think about why you are doing what you are doing? And then explain why you need this?

I understand that you as a user ask for this solution.
But I as a maintainer also have the knowledge that this solution isn’t a good idea.

I think there’s an actual problem you are running into, which needs a solution, but this solution isn’t it.

This is called the XY problem: https://github.com/rehypejs/.github/blob/main/support.md. See that link for some tips. And also check with a rubber duck, that often helps!

**Comment by ChristianMurphy** [comment]
https://github.com/rehypejs/rehype/issues/109#issuecomment-1280630895

Will be relevant here.
XAST and HAST are two different ASTs.

New parsers and ecosystems are welcome!
Though I'd tend to agree with wooorm, that shoving an different AST into rehype may not be the best fit.

And if it is intended to be another HAST/HTML parser, it wouldn't resolve #109.
And it would likely be its own project, much like https://github.com/rehypejs/rehype-dom

**Comment by milahu** [comment]
this is just a quick fix to convert xml (docbook) to markdown
"works for me" so far ...
but i dont use features like source locations, which have different props in the AST

idea is based on [cheerio](https://github.com/cheeriojs/cheerio), which also uses these 2 parsers

> shoving an different AST into rehype may not be the best fit

yeah ...
let me patch htmlparser2 to produce the same ast as parse5?

> I understand that you as a user ask for this solution.

ideally i want to reuse rehype, as html and xml only differ in syntax, but (can) have the same AST

**Comment by wooorm** [comment]
Thanks for sharing more!

Turning XML into markdown is not solved by this. It’s not solved by any tool we can make that works for everyone. XML is domain-specific, like JSON. You have a custom flavor of markdown that is not going to work for everyone that uses XML.

> as html and xml only differ in syntax

This is close but incorrect: XML is like JSON, it can represent anything!
You don’t have XML, you have XML, you have docbook.

What you need is not this, but something that turns docbook into mdast or hast or so.
That’s not solved here. You need different tools.

If you wanted a `rehype-parse` alternative that supports XML, here is it. But not that it isn’t rehype at all. It’s an XML parser:

```js
import {fromXml} from 'xast-util-from-xml'

export default function rexmlParse() {
  this.Parser = (value) => fromXml(value)
}
```

Then you need something else. Something that turns docbook into mdast or hast.
You need to write that yourself. Probably like [`hast-util-to-mdast`](https://github.com/syntax-tree/hast-util-to-mdast). I recommend converting docbook XML into hast though, and then later using `hast-util-to-mdast` itself.


> also, this allows to consume both html and xhtml in rehype


This project supports XHTML already. XHTML is a syntax format for HTML. The semantics of elements are still the same. The doctype switches between them. And that’s supported. You have a domain-specific XML format.

> btw, hows does rehype parse svg? (or mdx?)

rehype parses SVG in the HTML space. Just like a browser does. That means that it parser it according to the rules of HTML, not according to the rules of XML. Those are different, but in the case of SVG probably fine.
rehype **does not parse MDX**.

**Comment by wooorm** [comment]
Feel free to discuss here some more, but Discussions are the best place to get help.
Closing, because this solution does not work well for these packages. They can be somewhere else

**Comment by milahu** [comment]
> > also, this allows to consume both html and xhtml in rehype
> 
> This project supports XHTML already.

no! parse5 can not parse xhtml, and probably the svg parse tree has subtle bugs too

why would a project like cheerio use a different parser for xhtml? for fun ...?

example: in the htmlparser2 tree, xml namespaces are propagated to child nodes (not in the parse5 tree)

test:

<details>

https://astexplorer.net/ &rarr; html &rarr; parse5 vs htmlparser2

```xml
<?xml version="1.0"?>

<!--
https://www.w3.org/QA/2002/04/valid-dtd-list.html

trying to embed docbook xml into xhtml ... still invalid
https://validator.w3.org/check

but: parsing CDATA
parse5 fails
htmlparser2 succeeds
-->

<!DOCTYPE html PUBLIC
    "-//W3C//DTD XHTML 1.1 plus MathML 2.0 plus SVG 1.1//EN"
    "http://www.w3.org/2002/04/xhtml-math-svg/xhtml-math-svg.dtd">

<html
  xmlns="http://www.w3.org/1999/xhtml"
  xmlns:docbook="http://docbook.org/ns/docbook"
  xmlns:xlink="http://www.w3.org/1999/xlink"
  xmlns:xi="http://www.w3.org/2001/XInclude"
  xml:id="sec-functions-library-attrset"
  lang="en"
  xml:lang="en"
>

<head>
 <title>test</title>
</head>

<body>

<example xml:id="function-library-lib.attrset.attrByPath-example-value-exists">
<title>Extracting a value from a nested attribute set</title>
<programlisting><![CDATA[
let set = { a = { b = 3; }; };
in lib.attrsets.attrByPath [ "a" "b" ] 0 set
=> 3
]]></programlisting>
</example>

</body>

</html>
```

parse5:

```
<![CDATA[
let set = { a = { b = 3; }; };
in lib.attrsets.attrByPath [ "a" "b" ] 0 set
=
```

and

```
 3
]]>
```

htmlparser2:

```
<![CDATA[
let set = { a = { b = 3; }; };
in lib.attrsets.attrByPath [ "a" "b" ] 0 set
=> 3
]]>
```

</details>

ninja edit:

> > shoving an different AST into rehype may not be the best fit
> 
> yeah ... let me patch htmlparser2 to produce the same ast as parse5?
> 
> edit: this problem is solved in [htmlparser2-tree-adapter](https://github.com/inikulin/parse5/tree/master/packages/parse5-htmlparser2-tree-adapter)

**Comment by wooorm** [comment]
> no! parse5 can not parse xhtml, and probably the svg parse tree has subtle bugs too

Yes it can. You are mixing up things. XML is different from XHTML. Open that file in a browser, and it will do exactly what we and parse5 do.

`htmlparser2` is not an HTML parser. It’s not even an XML parser. It’s build to accept broken markup.

> why would a project like cheerio use a different parser for xhtml? for fun ...?

Because it doesn’t support HTML, XHTML, or XML. It supports a weird lax buggy format.

---

@milahu We’ve sent a bunch of tips and asked a bunch of questions. You ignored basically all of that. I’m not happy about helping you and then you ignoring that help...

**Comment by milahu** [comment]
> Open that file in a browser ...

good idea

> ... and it will do exactly what we and parse5 do.

no

in chrome:
cdata is parsed as one tag (cdata-section) ... not as 2 tags (comment and text) like by parse5

Screenshot_20221018_092101

**Comment by wooorm** [comment]
Can’t reproduce...

As `index.html`:

```
<?xml version="1.0"?>

<!--
https://www.w3.org/QA/2002/04/valid-dtd-list.html

trying to embed docbook xml into xhtml ... still invalid
https://validator.w3.org/check

but: parsing CDATA
parse5 fails
htmlparser2 succeeds
-->

<!DOCTYPE html PUBLIC
    "-//W3C//DTD XHTML 1.1 plus MathML 2.0 plus SVG 1.1//EN"
    "http://www.w3.org/2002/04/xhtml-math-svg/xhtml-math-svg.dtd">

<html
  xmlns="http://www.w3.org/1999/xhtml"
  xmlns:docbook="http://docbook.org/ns/docbook"
  xmlns:xlink="http://www.w3.org/1999/xlink"
  xmlns:xi="http://www.w3.org/2001/XInclude"
  xml:id="sec-functions-library-attrset"
  lang="en"
  xml:lang="en"
>

<head>
 <title>test</title>
</head>

<body>

<example xml:id="function-library-lib.attrset.attrByPath-example-value-exists">
<title>Extracting a value from a nested attribute set</title>
<programlisting><![CDATA[
let set = { a = { b = 3; }; };
in lib.attrsets.attrByPath [ "a" "b" ] 0 set
=> 3
]]></programlisting>
</example>

</body>

</html>
```

Chrome:

Screen Shot 2022-10-18 at 09 33 42


Safari:

Screen Shot 2022-10-18 at 09 34 07

Firefox:

Screen Shot 2022-10-18 at 09 35 27

**Comment by milahu** [comment]
> As `index.html`

sorry, gotta be `index.xhtml`

**Comment by wooorm** [comment]
So that’s different. Based on an XML mimetype the browser switches to an XML parser. See the HTML spec for more info.

What remains is that we have shown you how to solve your problem and you ignored it.
[Here I described](https://github.com/rehypejs/rehype/pull/112#issuecomment-1281298764) the 4 lines of code needed to create an XML parser for unified.
Then I explained how you can map your custom elements to mdast or hast, how you can turn docbook into markdown or HTML, which I’m sure other people will find useful.
I strongly believe that that is the solution.

**Comment by milahu** [comment]
> [Here I described](https://github.com/rehypejs/rehype/pull/112#issuecomment-1281298764) the 4 lines of code needed to create an XML parser for unified.

thanks, will use

> This project supports XHTML already.

this claim is still wrong

please take some time to meditate on this issue, maybe ask your rubber duck ...

until then, thanks for the discussion

**Comment by ChristianMurphy** [comment]
> > This project supports XHTML already.
> this claim is still wrong

XHTML is also valid HTML.
HTML can be parsed with parse5.

The main difference in XHTML is the removal of recovery modes which HTML includes.
If you need hard errors that XHTML provides, yes you'd need to use a proper XML parser, as xast does.

htmlparser2 is not a proper XML parser, it has even more recovery modes than HTML does.

**Comment by milahu** [comment]
i finally found a problem with htmlparser ...

from

```xml
<title><function>lib.attrset.attrByPath</function></title>
```

htmlparser will parse `<function>lib.attrset.attrByPath</function>` as text node

so yeah, back to

> [Here I described](https://github.com/rehypejs/rehype/pull/112#issuecomment-1281298764) the 4 lines of code needed to create an XML parser for unified.

thanks guys!

**Comment by milahu** [comment]
version 1 of docbook-to-markdown ... https://github.com/milahu/unifiedjs-docbook

currently, im going directly from xast to mdast (not via hast), so far so good.
probably the "via hast" route is better to embed html in markdown, for example `<dl>`

**Pull request closed without merging**